### PR TITLE
Open source is insecure, and other FUD

### LINKED ISSUE #66: FUD: Releasing open source is a security risk

### PR DESCRIPTION
**Pull request by benbalter**
@gbinal, @konklone any other common arguments to debunk?

Fixes #66.

### CODE REVIEW

File: _posts/2014-09-22-open-source-is-not-insecure.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+#### Anyone can see where the vulnerabilities are
+
+This is predicated, almost exclusively on the [(disproven to death) strategy of security through obscurity](http://en.wikipedia.org/wiki/Security_through_obscurity). It's a form of security theater. It's the idea of hiding a key under the welcome mat. It assumes that, if the bad guy doesn't know it's there, it's secure, right? Wrong. It's the same reason building blue prints can be available for public inspection (and city review), without increasing the chance of a break in. When everyone can see the inner workings (including those you want to keep out), you're forced to build inherently secure systems. Statistically, [open source software produces more secure software than its closed-source counterparts](http://www.infoworld.com/d/open-source-software/report-open-source-tops-proprietary-code-in-quality-187169). Think about it [this way](http://www.commlawblog.com/2010/01/articles/unlicensed-operations-and-emer/fcc-changes-stance-on-opensource-security/):
```
Comment (konklone) [inline _posts/2014-09-22-open-source-is-not-insecure.md]: Looks like this paragraph trailed off with more meant to be there?

File: _posts/2014-09-22-open-source-is-not-insecure.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+### Open source as a workflow
+
+The second form of open source FUD attacks open source as a workflow, meaning that by publishing the source code to your software, by that workflow's very nature, it some how, becomes less secure. Again, there are several flavors to this argument.
+
+#### Anyone can make changes
+
+The idea that anyone can make changes, is simply untrue — FUD at its best. I suspect this stems from using Wikipedia as an analogy to describe open source to the uninitiated. Simply put, it comes from a place of ignorance. You're not going to wake up one day and mysteriously, your code will be riddled with malware and backdoors.
```
Comment (konklone) [inline _posts/2014-09-22-open-source-is-not-insecure.md]: +:100: but I'd definitely hit the hammer after the first line, e.g.

> The idea that anyone can make changes is simply untrue -- FUD at its best. Like with closed projects, only authorized people can approve changes to open source projects.

File: _posts/2014-09-22-open-source-is-not-insecure.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+
+#### We'll expose our secrets
+
+Open or closed, industry-standard best practices dictate you shouldn't have secrets in your code. Passwords, tokens, server names, anything remotely secret sauce — heck, anything that might change based on the environment (development, staging, production) — should be placed in the database, an environmental variable, or in a VCS-ignored configuration file, not distributed with the software itself. For one, you've potentially got outside contractors or others in your code, which shouldn't be an access control mechanism in itself. For another, best practices once again dictate that such configurations aren't shared between environments, and are rolled frequently, something significantly harder if it requires a code change and deploy.
+
+Can open source be insecure? Of course. Is there insecure open source software? Of couse again. But the idea that a language, platform, or workflow, bassed solely on its philosophy, is someone inherently less secure is absurd, and I'd argue, can often be, if not exclusively, more secure,than closed source or proprietary alternatives when done right.
```
Comment (konklone) [inline _posts/2014-09-22-open-source-is-not-insecure.md]: `bassed` -> `based`

`more secure,than` -> `more secure, than`

File: _posts/2014-09-22-open-source-is-not-insecure.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+Not all software must be closed source and the decision whether a particular piece of software should be open or closed should be defined by industry-best practices, not by fear.
```
Comment (konklone) [inline _posts/2014-09-22-open-source-is-not-insecure.md]: `decision whether` -> `decision of whether`

File: _posts/2014-09-22-open-source-is-not-insecure.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+#### It's made by a bunch of hobbyists
+
+That's true of some projects, for sure. There's also fly-by-night software companies that sell sub-par, closed-source software. Again, quality is quality. Chances are, the project made in the developer's mom's basement isn't going up against the one made by the company that makes "enterprise-grade" solutions. Like anything else, [look to who's behind the software and who else uses it](http://ben.balter.com/2014/06/02/how-to-identify-a-strong-open-source-project/). Major projects like Linux, WordPress, and Drupal are relied on by thousands of for-profit business, and have "suits" supporting them, to ensure their success.
```
Comment (konklone) [inline _posts/2014-09-22-open-source-is-not-insecure.md]: This is good, but never quite gets around to saying that open source can be made by other than 1-person potentially-in-mom's-basement dudes. Suggestion:

> Major projects like Linux, WordPress, and Drupal are built by teams of professionals and hobbyists, relied on by thousands of for-profit business, and have "suits" supporting them, to ensure their success.

File: _posts/2014-09-22-open-source-is-not-insecure.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+excerpt: "The idea that open source software is inherently less secure than its closed source or proprietary counterparts is untrue and stems from fear, uncertainty, and doubt (FUD)."
+---
+
+Spend any time in the enterprise IT world, and you'll quickly learn that open source software is insecure, and only what you use when you can't afford "real" software — both false. This is a form of [FUD (fear, uncertainly, doubt)](http://en.wikipedia.org/wiki/Fear,_uncertainty_and_doubt), which has been historically encouraged [by those who sell proprietary alternatives](https://en.wikipedia.org/wiki/Halloween_documents).
+
+When talking about the security of open source, there's two, parallel conversations. Open source as a platform, and open source as a workflow.
+
+### Open source as a platform
+
+The first threshold issue, is whether building or using open source software, even if your own particular code isn't publishes, is some how inherently less secure. There are a few flavors to this argument.
```
Comment (konklone) [inline _posts/2014-09-22-open-source-is-not-insecure.md]: `some how` -> `somehow`

File: _posts/2014-09-22-open-source-is-not-insecure.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+That said, there is something to be said for a distributed workflow. Lots of closed-source solutions, influenced by nine-to-five, Monday-to-Friday workflows, stringent release cycles, and press paralysis [are forced to delay releases that patch critical vulnerabilities](http://www.xconomy.com/boston/2008/03/07/delays-in-software-patch-pushed-security-firm-to-disclose-vmware-flaw/), often for years, while open source projects, guaranteed to have someone awake at any hour of the night, can turn around that same patch, [literally within hours of the vulnerability being reported](https://twitter.com/#!/nacin/status/9753986051604480).
+
+### Open source as a workflow
+
+The second form of open source FUD attacks open source as a workflow, meaning that by publishing the source code to your software, by that workflow's very nature, it some how, becomes less secure. Again, there are several flavors to this argument.
```
Comment (konklone) [inline _posts/2014-09-22-open-source-is-not-insecure.md]: `some how,` -> `somehow`

**Comment by konklone** [comment]
Oh man, yes -- the more posts like this in the world, the better. Great job, @benbalter.

I left a few inline comments, but I do have one overarching suggestion, which is to avoid the cognitive trap of headlining the myths in myth form. In other words, saying "It's made by a bunch of hobbyists" and then explaining how it's not made by a bunch of hobbyists still leaves the brain kind of thinking that it's made by a bunch of hobbyists. (It's also inconsistent with other headlines that say the facts, like "Anyone can see where the vulnerabilities are", potentially confusing the reader.)

This effect is most potent with headlines, but the opening sentence has one too: "Spend any time in the enterprise IT world, and you'll quickly learn that open source software is insecure". I'd love a reword that doesn't introduce the article by presenting it as fact before revealing that it's faux fact.

Great timing, too -- I'm actually helping out a presentation by @RobertLRead on open source tomorrow, which cites some of your earlier work. I'll see if I can find an opportunity to tell people to read your next post. :)

**Comment by gbinal** [comment]
Looks great to me

**Comment by benbalter** [comment]
@konklone @gbinal thanks as always for the :eyes: and :thought_balloon:. :smile:

**Comment by konklone** [comment]
Really nice work :D

**Pull request merged**